core/src/main/scala/filodb.core/memstore/ratelimit/RocksDbCardinalityStore.scala

```diff
@@ -144,108 +160,129 @@ class RocksDbCardinalityStore(ref: DatasetRef, shard: Int) extends CardinalitySt
-   * ""{NotLastKeySeparator}"myWs1"{NotLastKeySeparator}"myNs11"{LastKeySeparator}"network_usage"
+   * 0
+   * 1{KeySeparator}"myWs1"
+   * 1{KeySeparator}"myWs2"
+   * 2{KeySeparator}"myWs1"{KeySeparator}"myNs11"
+   * 2{KeySeparator}"myWs1"{KeySeparator}"myNs12"
+   * 2{KeySeparator}"myWs2"{KeySeparator}"myNs21"
+   * 2{KeySeparator}"myWs2"{KeySeparator}"myNs22"
+   * 3{KeySeparator}"myWs1"{KeySeparator}"myNs11"{KeySeparator}"heap_usage"
+   * 3{KeySeparator}"myWs1"{KeySeparator}"myNs11"{KeySeparator}"cpu_usage"
+   * 3{KeySeparator}"myWs2"{KeySeparator}"myNs21"{KeySeparator}"network_usage"
    * </pre>
    *
-   * In the above tree, we simply do a prefix search on <pre> ""{NotLastKeySeparator}"myWs1"{LastKeySeparator} </pre>
+   * In the above tree, we simply do a prefix search on <pre> 2{KeySeparator}"myWs1"{KeySeparator} </pre>
    * to get namespaces under workspace myWs1.
    *
    * @param shardKeyPrefix Zero or more elements that make up shard key prefix
-   * @param prefixSearch If true, returns key that can be used to perform prefix search to
-   *                     fetch immediate children in trie. Use false to fetch one specific node.
+   * @param keyDepth size to append at the beginning of the key
+   *   Note: When keyDepth > shardKeyPrefix.size, builds keys as:
+   *         <keyDepth>{KeySeparator}<name-1>{KeySeparator}<name-2> ... <name-n>{KeySeparator}
+   *         When keyDepth == shardKeyPrefix.size, the final KeySeparator is omitted.
    * @return string key to use to perform reads and writes of entries into RocksDB
    */
-  private def toStringKey(shardKeyPrefix: Seq[String], prefixSearch: Boolean): String = {
+  private def toStringKeyPrefix(shardKeyPrefix: Seq[String], keyDepth: Int): String = {
     import RocksDbCardinalityStore._
-    if (shardKeyPrefix.isEmpty) {
-      if (prefixSearch) LastKeySeparator.toString else ""
-    } else {
-      val b = new StringBuilder
-      cforRange { 0 until shardKeyPrefix.length - 1 } { i =>
-        b.append(NotLastKeySeparator)
-        b.append(shardKeyPrefix(i))
-      }
-      if (prefixSearch) {
-        b.append(NotLastKeySeparator)
-        b.append(shardKeyPrefix.last)
-        b.append(LastKeySeparator)
-      } else {
-        b.append(LastKeySeparator)
-        b.append(shardKeyPrefix.last)
-      }
-      b.toString()
+
+    val b = new StringBuilder
+    b.append(keyDepth)
+    cforRange { 0 until shardKeyPrefix.length } { i =>
+      b.append(KeySeparator)
+      b.append(shardKeyPrefix(i))
+    }
+    if (keyDepth > shardKeyPrefix.size) {
+      b.append(KeySeparator)
     }
+    b.toString()
   }
 
-  private def cardinalityToBytes(card: Cardinality): Array[Byte] = {
+  /**
+   * Builds keys as:
+   *   <keyDepth>{KeySeparator}<name-1>{KeySeparator}<name-2> ... <name-n>
+   * The final {KeySeparator} is always omitted.
+   */
+  private def toStringKey(shardKeyPrefix: Seq[String]): String = {
```

Comment:
Document parameter as: fully specified RocksDB key. Result cannot be used for prefix search. It is used to lookup a single key.

Reply:
Done!